File: python/sparktk/jvm/jconvert.py
# vim: set encoding=utf-8

#  Copyright (c) 2016 Intel Corporation 
#
#  Licensed under the Apache License, Version 2.0 (the "License");
#  you may not use this file except in compliance with the License.
#  You may obtain a copy of the License at
#
#       http://www.apache.org/licenses/LICENSE-2.0
#
#  Unless required by applicable law or agreed to in writing, software
#  distributed under the License is distributed on an "AS IS" BASIS,
#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#  See the License for the specific language governing permissions and
#  limitations under the License.
#

from sparktk.dtypes import dtypes


class JConvert(object):
    """Handles misc. conversions going to and from Scala for simple non-primitive types, like lists, option, etc."""

    def __init__(self, jutils):
        self.jutils = jutils
        self.sc = jutils.sc
        self.scala = self.sc._jvm.org.trustedanalytics.sparktk.jvm.JConvert

    def list_to_double_list(self, python_list):
        return [float(item) for item in python_list]

    def to_scala_list_double(self, python_list):
        return self.scala.toScalaList(self.list_to_double_list(python_list))

    def to_scala_list(self, python_list):
        return self.scala.toScalaList(python_list)

    def to_scala_list_string(self, python_list):
        return self.scala.toScalaList([unicode(item) for item in python_list])

    def to_scala_list_string_bool_tuple(self, python_list):
        return self.scala.toScalaList([self.scala.toScalaTuple2(unicode(item[0]), item[1]) for item in python_list])

    def to_scala_list_string_option_tuple(self, python_list):
        return self.scala.toScalaList([self.scala.toScalaTuple2(unicode(item[0]), self.scala.toOption(item[1])) for item in python_list])

    def to_scala_vector_double(self, python_list):
        return self.scala.toScalaVector(self.list_to_double_list(python_list))

    def to_scala_vector_string(self, python_list):
        return self.scala.toScalaVector([unicode(item) for item in python_list])

    def to_scala_map(self, python_dict):
        return self.scala.toScalaMap(python_dict)

    def scala_map_to_python_with_iterable_values(self, m):
        result = self.scala_map_to_python(m)
        python_map_with_iterable_values = {}
        for k,v in result.items():
            python_map_with_iterable_values[k] = list(self.scala.scalaSeqToPython(v))
        return python_map_with_iterable_values

    def scala_map_to_python(self, m):
        return self.scala.scalaMapToPython(m)

    def scala_option_map_to_python(self, m):
        python_option = self.from_scala_option(m)
        return None if not python_option else self.scala_map_to_python(python_option)

    def to_scala_option_map(self, m):
        if isinstance(m, dict):
            scala_map = self.to_scala_map(m)
        elif m is None:
            scala_map = None
        else:
            raise ValueError("Expected a dict object, but received: %s" % type(m))
        return self.scala.toOption(scala_map)

    def to_scala_option(self, item):
        return self.scala.toOption(item)

    def to_scala_option_list_double(self, python_list):
        if isinstance(python_list, list):
            python_list = self.list_to_double_list(python_list)
        return self.to_scala_option(python_list)

    def to_scala_option_list_string(self, python_list):
        if isinstance(python_list, list):
            python_list = self.to_scala_list_string(python_list)
        return self.to_scala_option(python_list)

    def to_scala_option_either_string_int(self, item):
        if item is not None:
            return self.scala.toOption(self.scala.toEitherStringInt(item))
        else:
            return self.scala.toOption(item)

    def to_scala_date_time_list(self, python_list):
        return self.scala.toScalaList([self.scala.toDateTime(item) for item in python_list])

    def to_scala_date_time(self, item):
        return self.scala.toDateTime(item)

    def from_scala_option(self, item):
        return self.scala.fromOption(item)

    def from_scala_seq(self, seq):
        return self.scala.scalaSeqToPython(seq)

    def from_scala_vector(self, vector):
        return list(self.scala.scalaVectorToPython(vector))

    def from_scala_matrix(self, matrix):
        result = self.scala.scalaMatrixToPython(matrix)
        python_list = list(result)
        return [list(py_list) for py_list in python_list]

    def to_scala_group_by_aggregation_args(self, python_map):
        scala_map = self.to_scala_map(python_map)
        return self.sc._jvm.org.trustedanalytics.sparktk.frame.internal.ops.groupby.GroupByAggregationArgs(scala_map)

    def combine_scala_maps(self, python_list_of_scala_maps):
        scala_list = self.to_scala_list(python_list_of_scala_maps)
        return self.scala.combineScalaMap(scala_list)

File: sparktk-core/src/main/scala/org/trustedanalytics/sparktk/jvm/JConvert.scala
/**
 *  Copyright (c) 2016 Intel Corporation 
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */
package org.trustedanalytics.sparktk.jvm

import java.util
import java.util.{ ArrayList => JArrayList, List => JList, Map => JMap }

import org.apache.spark.SparkContext
import org.apache.spark.api.java.JavaSparkContext
import org.apache.spark.mllib.linalg.DenseMatrix
import org.joda.time.{ DateTime, DateTimeZone }
import org.apache.spark.org.trustedanalytics.sparktk.SparkAliases
import org.trustedanalytics.sparktk.frame.{ Column, DataTypes, FrameSchema }

import scala.collection.JavaConverters._
//import scala.collection.mutable

object JConvert extends Serializable {

  def toScalaSparkContext(jsc: JavaSparkContext): SparkContext = {
    JavaSparkContext.toSparkContext(jsc)
  }

  def toScalaList[T](x: JList[T]): List[T] = x.asScala.toList

  def toScalaVector[T](x: JList[T]): Vector[T] = x.asScala.toVector

  def toOption[T](item: T) = {
    item match {
      case null | None => None
      case item: JList[T] => Some(toScalaList(item))
      case _ => Some(item)
    }
  }

  def toEitherStringInt(item: Any): Either[String, Int] = {
    item match {
      case s: String => Left(s)
      case i: Int => Right(i)
    }
  }

  def toDateTime(item: Any): DateTime = {
    item match {
      case l: Long => new DateTime(l, DateTimeZone.UTC);
      case s: String => DateTime.parse(s)
      case _ => throw new IllegalArgumentException(s"Unable to translate type ${item.getClass.getName} to DateTime.")
    }
  }

  def fromOption(o: Option[Any]): Any = o.orNull

  def scalaMapToPython[K, V](m: Map[K, V]): JMap[K, V] = SparkAliases.JavaUtils.mapAsSerializableJavaMap(m)

  def scalaSeqToPython[T](seq: Seq[T]): JList[T] = seq.asJava

  def scalaVectorToPython[T](vector: Vector[T]): JList[T] = vector.asJava

  def scalaMatrixToPython(matrix: DenseMatrix): JList[JList[Double]] = {
    val (numRows, numCols) = (matrix.numRows, matrix.numCols)
    val result = new JArrayList[JArrayList[Double]]()
    for (i <- 0 until numRows) {
      val r = new JArrayList[Double]()
      for (j <- 0 until numCols) r.add(matrix(i, j))
      result.add(r)
    }
    result.asInstanceOf[JList[JList[Double]]]
  }

  def toScalaTuple2[T](a: T, b: T): (T, T) = (a, b)

  def toScalaMap[K, V](jm: java.util.Map[K, V]): Map[K, V] = SparkAliases.PythonUtils.toScalaMap(jm)

  def combineScalaMap[K, V](seq: Seq[Map[K, V]]): Map[K, V] = {
    seq.reduce(_ ++ _)
  }

  //  def frameSchemaToScala(pythonSchema: JArrayList[JArrayList[String]]): FrameSchema = {
  //    val columns = pythonSchema.asScala.map { item =>
  //      val list = item.asScala.toList
  //      require(list.length == 2, "Schema entries must be tuples of size 2 (name, dtype)")
  //      Column(list.head, DataTypes.toDataType(list(1)))
  //    }.toVector
  //    FrameSchema(columns)
  //  }
  //
  //    def frameSchemaToPython(scalaSchema: FrameSchema): JArrayList[JArrayList[String]] = {
  //      val pythonSchema = new JArrayList[JArrayList[String]]()
  //      scalaSchema.columns.map {
  //        case Column(name, dtype) =>
  //          val c = new JArrayList[String]()
  //          c.add(name)
  //          c.add(dtype.toString)
  //          pythonSchema.add(c)
  //      }
  //      pythonSchema
  //    }

  /**
   * Convert an RDD of Java objects to an RDD of serialized Python objects, that is usable by
   * PySpark.
   *
   * (private function taken from Spark's org.apache.spark.api.python.SerDeUtil)
   */
  //  def javaToPython(jrdd: JavaRDD[Row]): JavaRDD[Array[Byte]] = {
  //    //def javaToPython(jRDD: JavaRDD[_]): JavaRDD[Array[Byte]] = {
  //    scalaToPython(jrdd.rdd) //.mapPartitions { iter => new AutoBatchedPickler(iter) }
  //  }

  //  def scalaToPython(rdd: RDD[Row], schema: Schema): JavaRDD[Array[Byte]] = {
  //    //def javaToPython(jRDD: JavaRDD[_]): JavaRDD[Array[Byte]] = {
  //    println("In scalaToPython...")
  //    //SerDeUtil.javaToPython(rdd.toJavaRDD())
  //
  //    //# these two lines work, but rows show up as tuples...
  //    val raa: RDD[Any] = toAnyRDD(rdd, schema) // , rdd)
  //    EvaluatePython.javaToPython(raa)
  //
  //    // don't know about these two lines:
  //    //rdd.mapPartitions { iter => EvaluatePython.rowToArray}
  //    //rdd.mapPartitions { iter => new AutoBatchedPickler(iter) }
  //  }
  //
  //  def toAnyRDD(rdd: RDD[Row], schema: Schema): RDD[Any] = {
  //    val anyRDD: RDD[Any] = rdd.map(row => {
  //      //      val rowArray = new Array[Any](row.length)
  //      //      row.toSeq.zipWithIndex.map {
  //      //        case (o, i) =>
  //      //          o match {
  //      //            case null => null
  //      //            case _ => rowArray(i) = o
  //      //            //              val colType = schema.column(i).dataType
  //      //            //              val dType: DataType = FrameRdd.schemaDataTypeToSqlDataType(colType)
  //      //            //              try {
  //      //            //                rowArray(i) = EvaluatePython.fromJava(o, dType)
  //      //            //              }
  //      //            //              catch {
  //      //            //                case e: Exception => null
  //      //            //              }
  //      //          }
  //      //      }.toArray
  //      //      rowArray
  //      row.toSeq.toArray
  //      //new GenericRow(rowArray)
  //    })
  //    anyRDD
  //  }
  //
  //  def pythonToScala(jrdd: JavaRDD[Array[Byte]], schema: Schema): RDD[Row] = {
  //    val raa: JavaRDD[Array[Any]] = pythonToJava(jrdd)
  //    toRowRdd(raa.rdd, schema)
  //  }
  //
  //  def pythonToJava(jrdd: JavaRDD[Array[Byte]]): JavaRDD[Array[Any]] = {
  //    val j = SerDeUtil.pythonToJava(jrdd, batched = true)
  //    SerDeUtil.toJavaArrayAny(j)
  //  }
  //
  //  def toRowRdd(raa: RDD[Array[Any]], schema: Schema): RDD[org.apache.spark.sql.Row] = {
  //    val rowRDD: RDD[org.apache.spark.sql.Row] = raa.map(row => {
  //      val rowArray = new Array[Any](row.length)
  //      row.zipWithIndex.map {
  //        case (o, i) =>
  //          o match {
  //            case null => null
  //            case _ =>
  //              val colType = schema.column(i).dataType
  //              try {
  //                rowArray(i) = colType.parse(o).get
  //              }
  //              catch {
  //                case e: Exception => null
  //              }
  //          }
  //      }
  //      new GenericRow(rowArray)
  //    })
  //    rowRDD
  //  }
}
